tests/test_browser.go

```diff
@@ -0,0 +1,297 @@
+func (b *browser) withFileServer() *browser {
+	const (
+		slash = string(os.PathSeparator)
+		path  = slash + testBrowserStaticDir + slash
+	)
+
+	fs := http.FileServer(http.Dir(testBrowserStaticDir))
+
+	return b.withHandle(path, http.StripPrefix(path, fs).ServeHTTP)
+}
+
+// URL returns the listening HTTP test server's URL combined with the given path.
+func (b *browser) URL(path string) string {
+	if b.http == nil {
+		panic("You should enable HTTP test server, see: withHTTPServer option")
+	}
+	return b.http.ServerHTTP.URL + path
+}
+
+// staticURL is a helper for URL("/`testBrowserStaticDir`/"+ path).
+func (b *browser) staticURL(path string) string {
+	return b.URL("/" + testBrowserStaticDir + "/" + path)
+}
+
+// attachFrame attaches the frame to the page and returns it.
+func (b *browser) attachFrame(page api.Page, frameID string, url string) api.Frame {
+	pageFn := `
+	async (frameId, url) => {
+		const frame = document.createElement('iframe');
+		frame.src = url;
+		frame.id = frameId;
+		document.body.appendChild(frame);
+		await new Promise(x => frame.onload = x);
+		return frame;
+	}
+	`
+	return page.EvaluateHandle(
+		b.rt.ToValue(pageFn),
+		b.rt.ToValue(frameID),
+		b.rt.ToValue(url)).
+		AsElement().
+		ContentFrame()
+}
+
+// detachFrame detaches the frame from the page.
+// func (b *browser) detachFrame(page api.Page, frameID string) {
+// 	pageFn := `
+// 	frameId => {
+//         	document.getElementById(frameId).remove();
+//     	}
+// 	`
+// 	page.Evaluate(
+// 		b.rt.ToValue(pageFn),
+// 		b.rt.ToValue(frameID))
+// }
+
+// launchOptions provides a way to customize browser type
+// launch options in tests.
+type launchOptions struct {
+	Debug    bool   `js:"debug"`
+	Headless bool   `js:"headless"`
+	SlowMo   string `js:"slowMo"`
+	Timeout  string `js:"timeout"`
+}
+
+// withLaunchOptions is a helper for increasing readability
+// in tests while customizing the browser type launch options.
+//
+// example:
+//
+//    b := TestBrowser(t, withLaunchOptions{
+//        SlowMo:  "100s",
+//        Timeout: "30s",
+//    })
+type withLaunchOptions = launchOptions
+
+// defaultLaunchOptions returns defaults for browser type launch options.
+// TestBrowser uses this for launching a browser type by default.
+func defaultLaunchOpts() launchOptions {
+	var (
+		debug    = false
+		headless = true
+	)
+	if v, found := os.LookupEnv("XK6_BROWSER_TEST_DEBUG"); found {
+		debug, _ = strconv.ParseBool(v)
+	}
+	if v, found := os.LookupEnv("XK6_BROWSER_TEST_HEADLESS"); found {
+		headless, _ = strconv.ParseBool(v)
+	}
+
+	return launchOptions{
+		Debug:    debug,
+		Headless: headless,
+		SlowMo:   "0s",
+		Timeout:  "30s",
+	}
+}
+
+// httpServerOption is used to detect whether to enable the HTTP test
+// server.
+type httpServerOption struct{}
+
+// withHTTPServer enables the HTTP test server.
+//
+// example:
+//
+//    b := TestBrowser(t, withHTTPServer())
+func withHTTPServer() httpServerOption {
+	return struct{}{}
+}
```

Comment:
Hhmm do we need this and `fileServerOption` to be separate types? The way this is defined and checked is a bit convoluted. Wouldn't it be simpler if they're both just boolean fields of a `browserConfig` struct and we allow that to be passed directly to the constructor?

Reply:
They serve different things. You want to have an HTTP server but not a file server. But a file server enables an HTTP server.

Config struct is perpendicular to the current design I established here. I intentionally tried to avoid that. The current one is concise (_one of the goals in Go_), and doesn't require us to pass default or nil configs while getting a test browser, and provides us an ability to extend it with a natural language.

The premier design goal of the test browser is hiding complexity behind a simple interface—as Go does. So, it doesn't matter whether the implementation is more straightforward. More important is whether it's easy to use, readable, clear, concise (**for making a test's intentions pop up**), and prevents possible mistakes.

Passing a config struct is the opposite of most of these things: Harder to use, less readable, unclear, verbose, and makes a test's intentions fade out behind a test server's configuration. It also invites mistakes. See [this](https://www.youtube.com/watch?v=ltqV6pDKZD8)

I'm happy with the current design, but maybe we can do something like [this](https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/text/template/option.go).

Reply:
OK, thanks for explaining. I think this is overengineered for a test package and I probably wouldn't even consider this approach until it actually becomes a problem, but you feel strongly about it, and I don't disagree, so let's leave it as is.